cocos2d/CCImage.h

```diff
@@ -0,0 +1,149 @@
+/*
+ * cocos2d for iPhone: http://www.cocos2d-iphone.org
+ *
+ * Copyright (c) 2014 Cocos2D Authors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+
+#import <CoreGraphics/CGImage.h>
+
+#import "ccTypes.h"
+
+
+@class CCFile;
+
+
+/**
+ How much to rescale the image while loading it.
+ 
+ The default value is 1.0.
+ 
+ @warning Some image loaders may only support inverse powers of two (1/2, 1/4, etc)
+
+ @since 4.0
+ */
+extern NSString * const CCImageOptionRescaleFactor;
+
+/**
+ Should the image dimensions be expanded to a power of two while loading?
+ 
+ The default value is `NO`.
+
+ @since 4.0
+ */
+extern NSString * const CCImageOptionExpandToPOT;
+
+/**
+ Should the image alpha be premultiplied while loading?
+ 
+ The default value is YES.
+ 
+ @warning Some image loaders only support pre-multiplied alpha.
+
+ @since 4.0
+ */
+extern NSString * const CCImageOptionPremultiply;
+
+/**
+ Low level image handling class for RGBA8 bitmaps.
+ */
+@interface CCImage : NSObject<NSCoding>
+
+/**
+ Initialize a new image with raw pixel data. All default options are applied to the image.
+ 
+ TODO Should this initializer support an options dictionary?
+ TODO Should `pixelData` be allowed to be `nil`?
+
+ @param pixelSize    Size of the image in pixels.
+ @param contentScale Content scale of the image.
+ @param pixelData    A pointer to raw, tightly packed, RGBA8 pixel data.
+
+ @return An image object that wraps the given pixel data.
+
+ @since 4.0
+ */
+-(instancetype)initWithPixelSize:(CGSize)pixelSize contentScale:(CGFloat)contentScale pixelData:(NSMutableData *)pixelData;
```

Comment:
I waffled on that several times, but maybe it doesn't make sense. The original proposal used NSData.

@Birkemose's main suggestion was to make it possible to, for example, allow calculating normal maps at load time from a heightmap. I had been thinking about similar ideas about loading filters in general, so this fit in well. I changed it to mutable at some point so that it wouldn't require a copy to filter the image data.

On the other hand, very very few interesting image filters can be performed in place without a copy anyway. Also, since textures are uploaded on the rendering thread, this opens up the possibility for race conditions if the multi-threaded renderer is used. So... maybe mutable was a bad decision after all. It only saved one line of code for people applying filters to loaded image data, which is probably a very small number of people anyway.